### PR TITLE
`retrieve-state` tool fanout to peers

### PR DESCRIPTION
**Pull request by dwerner**
This PR introduces some parallel downloading of tries in the `retrieve-state` tool, which reduces it's runtime significantly from about 10 hours to approximately 15 mins in my own case.

### CODE REVIEW

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/lib.rs]: Did you want to leave this comment in?

File: execution_engine/src/core/engine_state/mod.rs
Comment (xcthulhu) [inline execution_engine/src/core/engine_state/mod.rs]: This function does not appear to be used...

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/lib.rs]:
```suggestion
```

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/lib.rs]: Tries or MB / sec might be helpful here...

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/lib.rs]: Bit of a show-stopper... I get some serious hiccups because of this

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/lib.rs:664]:
```suggestion
                    let bytes: Vec<u8> = blob.into();
                    if Digest::hash(bytes) != next_trie_key {
                        <...ban peer somehow...>
                        base_send
                        .send(Ok(PeerMsg::NoBytes {
                            trie_key: next_trie_key,
                            peer: address,
                        }))
                        .await
                        .expect("should send");
                    }
```

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [reply utils/retrieve-state/src/lib.rs]: Also, this is stalling for me... and it would be nice to have timestamps.

Maybe switch use logging like we have in the node/execution engine?

File: utils/retrieve-state/src/main.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/main.rs]: I thought we were using LMDB now?

It might be a little easier if we just had a `--data-dir`, and the subdirectories were the same as they are for the node.  I know a planned use case for this tool is to clone main-net for dry-run testing of upgrades...

File: utils/retrieve-state/src/main.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/main.rs]: I suggest we just do this by default rather than have a flag.

File: utils/retrieve-state/src/main.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/main.rs]: This isn't used...

File: utils/retrieve-state/src/main.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/main.rs:90]: It might be a good idea to have an option for how many workers. But we'd have to drop the `tokio::main` macro and use the builder in that case.

File: utils/retrieve-state/src/main.rs
Comment (afck) [inline utils/retrieve-state/src/main.rs]: Should either be "Specify" or "Specifies" everywhere.
Or maybe better: None of them? Just "the port to reach peer RPC endpoints at"?

File: utils/retrieve-state/src/main.rs
Comment (afck) [inline utils/retrieve-state/src/main.rs]:
```suggestion
            // also use the server we provided initially that peers were gathered from.
```

File: utils/retrieve-state/src/main.rs
Comment (afck) [inline utils/retrieve-state/src/main.rs]: We could also print the block hash here, so that a careful user can e.g. compare with a block explorer.

File: utils/retrieve-state/src/main.rs
Comment (afck) [inline utils/retrieve-state/src/main.rs]: Would it be better to configure that default with `structop` above instead?

File: utils/retrieve-state/src/main.rs
Comment (afck) [inline utils/retrieve-state/src/main.rs]: Would it be better to call this `automatic_sync` and make it just a `bool` (not an option) instead? Then omitting it would make it `false`, i.e. manual sync.

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:82]:
```suggestion
    rpc::rpc(client, url, "query_global_state", params).await
```

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:92]:
```suggestion
    rpc::rpc(client, url, "chain_get_era_info", params).await
```

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:122]: Why do we need that function? Shouldn't we always either specify a block or use the latest block?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:376]:
```suggestion
    /// Download and save the trie under this hash
```
?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:390]:
```suggestion
/// Downloads the trie from the peers to the provided LMDB path.
```

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:410]: Why not actually print that error here?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:476]: Not sure how likely that is in practice, but that could add a key for which a request is already in flight.

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:388]: Is this a single message enum for both directions? Aren't the variants sent in one direction disjoint from the ones sent in the other direction? Could this be two separate enums?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:427]: Why not also count the root right away, instead of adding `1` at the end?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:583]: Would it be a bit faster to first send to _all_ of them, and then await all of them?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:641]: Why not just a single stream, and a `Shutdown` message variant?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:658]: If every peer is only a single task and always immediately `await`s the result of `get_trie` before making the next request, are we really ever making multiple requests to the same peer at once, as suggested by `concurrent_requests_per_peer`? Or are we just putting `concurrent_requests_per_peer` items into the channel at once?

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:573]: Can this happen if `tries_downloaded == total_tries_count + 1`?

File: utils/retrieve-state/src/main.rs
Comment (xcthulhu) [reply utils/retrieve-state/src/main.rs]: I suggest that this code avoid trusted RPC calls and instead require a trusted hash.  It'd then get blocks to current using finality signatures to verify them.

@dwerner WDYT?

(might be good to leave in the current behavior under an `--unsafe` flag)

File: utils/retrieve-state/src/main.rs
Comment (dwerner) [reply utils/retrieve-state/src/main.rs]: This was a migration to load blocks from their old files into storage, and can be removed now.

File: execution_engine/src/core/engine_state/mod.rs
Comment (dwerner) [reply execution_engine/src/core/engine_state/mod.rs]: Thanks, I was trying out a mode where I decoupled the missing check, but this is unused now. Removed.

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs]: I was trying out spawn_blocking, but this actually led to some performance issues where we overuse the threadpool. Removed.

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs]: Removed

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs]: Stale comment, not related to the pauses we see where the trie is backed up on writes.

File: utils/retrieve-state/src/main.rs
Comment (dwerner) [reply utils/retrieve-state/src/main.rs]: Done

File: utils/retrieve-state/src/main.rs
Comment (dwerner) [reply utils/retrieve-state/src/main.rs]: You can do this now since the `highest_block` parameter is a `BlockIdentifier`. I'll add a note that asking for the highest block by height is considered insecure.

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs:427]: I'm using the tries downloaded to determine when the downloading has completed, whereas if I counted it right away, I would lose that resolution.

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs:641]: The idiom here is adapting a oneshot into a stream that is selected over with the response from the other stream. Warp uses this sort of api here https://docs.rs/warp/0.3.2/warp/struct.Server.html#method.serve_incoming_with_graceful_shutdown

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs:658]: Correct, this is badly named and instead can be thought of as the `peer`'s mailbox in an actor model sense. I've renamed it appropriately.

File: utils/retrieve-state/src/lib.rs
Comment (dwerner) [reply utils/retrieve-state/src/lib.rs:573]: Logically, no. `tries_downloaded == total_tries_count + 1` is used here as the completion check, but we still rely on `missing_trie_keys` to tell us if we've downloaded the _correct_ tries. This shouldn't be a problem, but I wanted a check in place.

File: utils/retrieve-state/src/lib.rs
Comment (afck) [reply utils/retrieve-state/src/lib.rs:658]: But then why do we have the `concurrent_requests_per_peer` setting at all?
I think we _do_ want to make that many requests at the same time, instead of `await`ing here immediately?

File: utils/retrieve-state/src/main.rs
Comment (xcthulhu) [reply utils/retrieve-state/src/main.rs]: I'd still prefer an `--unsafe` so they have to acknowledge they are doing something dangerous... like `kill -9` or `sudo`.

As discussed there is a security protocol for getting the highest block.  I'll see if I can quick, prototype something for you...

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [reply utils/retrieve-state/src/lib.rs:476]: Since the Merkle trie truly is a tree, I don't think that is possible @afck

File: utils/retrieve-state/src/lib.rs
Comment (afck) [reply utils/retrieve-state/src/lib.rs:658]: Is `peer_mailbox_size` really useful? I think we'd really want to make concurrent requests instead.

File: utils/retrieve-state/src/lib.rs
Comment (afck) [reply utils/retrieve-state/src/lib.rs:427]: Yes, I'm not suggesting adding the root to `tries_downloaded` right away, just to `total_tries_count`, so that the `tries_downloaded == total_tries_count + 1` check below doesn't need the `+ 1` anymore.

**Comment by xcthulhu** [comment]
@afck Can you run this tool locally?

```bash
cargo run --release --bin retrieve-state -- -a download-tries -n 3.14.161.135
```

@dwerner says it should take 5 minutes...

**Comment by goral09** [comment]
Let's extract [`d9e15bf` (#2353)](https://github.com/casper-network/casper-node/pull/2353/commits/d9e15bf8747aefdd9f311411e498492a6870e55c) from this PR to a separate one. It's not inherent to this ticket and should not be part of the changes. Extracting it to a separate PR will make it simpler to review, without having the `retrieve-state` tool context in mind.

**Comment by xcthulhu** [comment]
> It's not inherent to this ticket and should not be part of the changes.

While it could be a separate PR as you say, the tool doesn't work on my machine without this...

File: utils/retrieve-state/src/lib.rs
Comment (xcthulhu) [inline utils/retrieve-state/src/lib.rs]:
```suggestion
    Ok(())
```

**Comment by dwerner** [comment]
Update on this:

I've integrated @marc-casperlabs 's `work_queue` here now. I had to adapt it a bit but tried to stay true to what other parts are being used by fast-sync.

```
RUST_LOG=info cargo run --release --bin retrieve-state -- -a download-tries -n <node_ip_address>
```

File: utils/retrieve-state/src/lib.rs
Comment (afck) [inline utils/retrieve-state/src/lib.rs:878]: Why kill the worker in that case?

Comment (dwerner) [reply utils/retrieve-state/src/lib.rs:878]: In my tests, keeping the worker around led to a lot more attempts to request other tries from the same worker that failed with zero bytes.

Comment (afck) [reply utils/retrieve-state/src/lib.rs:878]: Ah, I just realized that each worker corresponds to one peer.
Isn't that inefficient? We should easily be able to make multiple parallel requests to the same peer.

Comment (dwerner) [reply utils/retrieve-state/src/lib.rs:878]: Have you tried running the tool?

File: resources/test/rest_schema_status.json
Comment (Fraser999) [inline resources/test/rest_schema_status.json:2]: It seems like the bulk of the change here is formatting, making spotting the actual diff fairly hard.

If the schema is generated by just using `serde_json::to_string_pretty` we can avoid the large diff.

File: execution_engine/src/core/engine_state/mod.rs
Comment (Fraser999) [inline execution_engine/src/core/engine_state/mod.rs:1793]: We should document exactly what effect the `check_integrity` flag has.

File: execution_engine/src/storage/global_state/in_memory.rs
Comment (Fraser999) [inline execution_engine/src/storage/global_state/in_memory.rs:272]: Let's document the new arg please.

File: execution_engine/src/storage/global_state/lmdb.rs
Comment (Fraser999) [inline execution_engine/src/storage/global_state/lmdb.rs:223]: Let's document the new arg please.

File: execution_engine/src/storage/global_state/mod.rs
Comment (Fraser999) [inline execution_engine/src/storage/global_state/mod.rs:118]: Let's document the new arg please.

File: execution_engine/src/storage/trie_store/operations/mod.rs
Comment (Fraser999) [inline execution_engine/src/storage/trie_store/operations/mod.rs:242]: The doc comment needs updated as point 2 is conditional now.

File: utils/retrieve-state/src/lib.rs
Comment (Fraser999) [inline utils/retrieve-state/src/lib.rs]: `kbps` is generally an acronym for kilobits per second.  Can we rename all "bps" output strings and variables to be explicit please?

**Review by Fraser999 — COMMENTED** [review]
The work queue is due to get unit tests added in the `feat-fast-sync` PR - let's incorporate that here too please.

The tool's `--help` command doesn't really explain all the various options very well.  I wouldn't block the PR for that, but it seems like this tool could get used a fair bit, and I'd really like better help output, or even a comprehensive README.

**Comment by dwerner** [comment]
> The work queue is due to get unit tests added in the feat-fast-sync PR - let's incorporate that here too please.

@marc-casperlabs could you please port over the unit tests here when they land in `feat-fast-sync`? Should be a simple copy into this branch (which is based on dev btw).

> The tool's --help command doesn't really explain all the various options very well. I wouldn't block the PR for that, but it seems like this tool could get used a fair bit, and I'd really like better help output, or even a comprehensive README.

I agree, but following this PR I'd like to move this to a separate repo, so I'd like to prioritize that work first. @Fraser999

File: resources/test/rest_schema_status.json
Comment (dwerner) [reply resources/test/rest_schema_status.json:2]: Done, hopefully it isn't too controversial that I added a file write of that `actual` value to disk, so that a quick copy can be performed, manually, once the diff is inspected.

File: execution_engine/src/storage/trie_store/operations/mod.rs
Comment (dwerner) [reply execution_engine/src/storage/trie_store/operations/mod.rs:242]: @xcthulhu can you once-over my doc comments here to see if more detail is needed? I believe this came in from your PR to my branch to improve the performance.

**Review by Fraser999 — APPROVED** [review]
Optimistic approval given that the WorkQueue tests are ported over.

I only reviewed the production code changes in depth, so it'd be worth waiting until you get approval from someone more familiar with the tool than me.  I'm also fine to leave the help/readme improvements until a subsequent PR.

**Comment by afck** [comment]
I tried it with testnet and get a throughput rate of ~260 kilobytes per second. I didn't wait for the full download to finish, though.

**Review by afck — APPROVED** [review]
Approving since none of my remaining comments are blockers.

**Pull request merged**
